docs/algorithms/nautilus/algorithm_en.md

```diff
@@ -0,0 +1,290 @@
+```c
+int res = express();
+if (res != 0 && res % 314 == 0) crash();
+```
+You have to instrument the target binary by AFL in Nautilus mode.
+
+The Nautilus mode of fuzzuf provides the following options:
+
+- `--out_dir`, `-o`: Path to folder to save the fuzzing result [**Required**]
+- `--exec_timelimit_ms`: Time limit of an execution (ms) [Default: 1000]
+- `--exec_memlimit`: Memory limit of the target binary (MB) [Default: 25]
+- `--grammar`: Path to grammar file [**Required**]
+- `--bitmap-size`: Bitmap size [Default: 1<<16]
+- `--generate-num`: The number of test cases generated in one fuzz loop [Default: 100]
+- `--detmut-num`: Number of cycles to execute deterministic mutations [Default: 1]
+- `--max-tree-size`: Maximum size of generated tree [Default: 1000]
+- `--forksrv`: Enable/disable fork server mode. Default to true. (not recommended to disable)
+
+You can fuzz the calculator like this, for example:
+```
+$ fuzzuf nautilus --out_dir output \
+                  --grammar ./calc_grammar.json \
+                  -- ./test/put_binaries/nautilus/calc @@
+```
+It is working successfully if a screen showing the status of fuzzing in real time appears.
+
+If your usage or grammar file is wrong, an error will show up.
+
+- `Grammar does not exist!`: Grammar file specified by `--grammar` does not exist.
+- `Unknown grammar type ('.json' expected)`: The extension of grammar file is not ".json"
+- `Cannot parse grammar file`: The content of grammar file is wrong. (Check your grammar file as explained in section 2-2.)
+
+## 3. Algorithm
+Generally, grammar based fuzzers generates AST (abstract syntax tree) according to a specific grammar and mutates a part of the AST to create testcases. Nautilus internally only uses the tree representation and mutates the tree.
+In this section, we explain the design of how Nautilus generates and mutates the testcases.
+
+### 3-1. Generation
+Since a non-terminal may have multiple rules, we need an algorithm to decide which rule to pick up. Nautilus uses uniform generation algorithm.
+Let's consider the following grammar:
+```
+<PROG> := <STMT>
+<PROG> := <STMT>; <PROG>
+<STMT> := return 1
+<STMT> := <VAR> = <EXPR>
+<VAR>  := a
+<EXPR> := <NUMBER>
+<EXPR> := <EXPR> + <EXPR>
+<NUMBER> := 1
+<NUMBER> := 2
+```
+For example, `<STMT>` has 2 rules: `return 1` or `<VAR> = <EXPR>`. If we choose a rule for each non-terminal by naive randomness, `return 1` is chosen with 50% probability. On the other hand, if we choose `<VAR> = <EXPR>`, `<EXPR>` has another 2 rules: `<NUMBER>` and `<EXPR> + <EXPR>`. The probability that one of them is chosen is 25% from `<STMT>`.
+The deeper part of the tree is selected with less probability with a naive randomness like this, which results in generating similar testcases. Naitlus, on the other hand, uses an algorithm by McKenzie[^2] so that it can select every rule in the grammar with the same probability.
```

Comment:
```suggestion
The deeper part of the tree is selected with less probability with naive randomness like this, which results in generating similar testcases. Nautilus, on the other hand, uses an algorithm by McKenzie[^2] so that it can select every rule in the grammar with the same probability.
```